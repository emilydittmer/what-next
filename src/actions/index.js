export const setSearchShow = show => ({
  type: 'SET_SEARCH_SHOW',
  show
})