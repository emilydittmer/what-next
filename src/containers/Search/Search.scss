.search-container {
  display: flex;
  align-items: center;
}

.search-field {
  top: 50%;
  left: 30%;
  position: fixed;
  padding: 0;
  margin: 0;
  height: 50px;
}

.search-input {
  height: 50px;
  width: 500px;
  font-size: 1.35em;
  font-family: 'Montserrat', Helvetica, sans-serif;
  font-weight: 300;
  padding: 0;
  padding-left: 10px;
  border-radius: 3%;
  margin: 0;
}

.submit-btn {
  height: 50px;
  width: 50px;
  border-width: 2px;
  margin:0;
  padding: 1px;
  font-family: 'Montserrat', Helvetica, sans-serif;
  background-color: hsla(4, 62%, 30%, 1);
  color: #fff;
}

.no-results {
  color: hsla(4, 62%, 30%, 1);
  font-size: 1.25em;
  font-weight: 700;
  text-shadow: -1px 0 white, 0 1px white, 1px 0 white, 0 -1px white;
}