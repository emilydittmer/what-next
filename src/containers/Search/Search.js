import React, { Component } from 'react';
import './Search.scss';
import { fetchWatchedShowId } from '../../utils/apiCalls';
import { setSearchShow } from '../../actions';
import { connect } from 'react-redux';

class Search extends Component {
  constructor(){
    super();
    this.state = {
      error: '',
      searchValue: ''
    }
  }

  handleChange= (event) => {
    event.preventDefault();
    this.setState({ [event.target.name]: event.target.value })
  }

  handleSubmit = async event => {
    event.preventDefault();

    const { searchValue }= this.state;
    let searchedShow = await fetchWatchedShowId(searchValue)
    if (searchedShow){
      this.props.setSearchShow(searchedShow)
    } else {
      this.setState({error: "No results found. Please search again"})
    }
    // this.findSuggestions()
    this.resetInputs();
  }

  // findSuggestions = async


  resetInputs = () => {
    this.setState({searchValue: ''})
  }

  render(){
    return(
      <form className='search-field'>
        <input 
          className='search-input'
          type='text'
          placeholder='TV Show Title'
          value={this.state.searchValue}
          name='searchValue'
          onChange={event => this.handleChange(event)}
        />
        <input
          className='submit-btn'
          type='submit'
          value= 'Submit'
          onClick = {event => this.handleSubmit(event)}
          />
      </form>
    )
  }
}


const mapDispatchToProps = dispatch => ({
  setSearchShow: show => dispatch(setSearchShow(show))
})

export default connect(null, mapDispatchToProps)(Search);