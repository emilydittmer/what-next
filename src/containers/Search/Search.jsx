import React, { Component } from "react";
import "./Search.scss";
import { fetchWatchedShowId } from "../../utils/apiCalls";
import { setSearchShow } from "../../actions";
import { connect } from "react-redux";

export class Search extends Component {
  constructor(props) {
    super(props);
    this.state = {
      error: "",
      searchValue: ""
    };
  }

  handleChange = event => {
    event.preventDefault();
    this.setState({ searchValue: event.target.value, error: '' });
  };

  handleSubmit = async event => {
    event.preventDefault();

    const { searchValue } = this.state;
    let searchedShow = await fetchWatchedShowId(searchValue);
    console.log(searchedShow);

    if (!searchedShow.error) {
      console.log(searchedShow);
      this.props.setSearchShow(searchedShow);
      this.props.history.push("/results");
    } else {
      console.log(searchedShow);
      this.setState({ error: "No results found. Please search again" });
    }
    this.resetInputs();
  };

  resetInputs = () => {
    this.setState({ searchValue: "" });
  };

  render() {
    return (
      <form className="search-field">
        <div className="search-container">
          <input
            className="search-input"
            type="text"
            placeholder="TV Show Title"
            value={this.state.searchValue}
            name="searchValue"
            onChange={event => this.handleChange(event)}
          />
          <input
            className="submit-btn"
            type="submit"
            value="Submit"
            onClick={event => this.handleSubmit(event)}
          />
        </div>
        {this.state.error === "" ? <p /> : <p>{this.state.error}</p>}
      </form>
    );
  }
}

export const mapDispatchToProps = dispatch => ({
  setSearchShow: show => dispatch(setSearchShow(show))
});

export default connect(
  null,
  mapDispatchToProps
)(Search);
