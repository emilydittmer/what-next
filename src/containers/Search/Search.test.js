import React from 'react';
import { shallow } from 'enzyme';
import { Search, mapDispatchToProps } from './Search';
import { setSearchShow } from "../../actions";

describe('Search', () => {
  let wrapper, props;
  let mockEvent = { 
    preventDefault: jest.fn(),
    target: {
    name: 'searchValue',
    value: 'The Office'
    }
  }

  beforeEach(() => {
    wrapper = shallow(<Search {...props} />)
  })

  it('should match the snapshot', () => {
    wrapper.setState({error: '', searchValue: ''})
    expect(wrapper).toMatchSnapshot()
  })

  it('should render an error', () => {
    wrapper.setState({error: 'Error', searchValue: ''})
    expect(wrapper).toMatchSnapshot()
  })

  it('should reset the searchValue state', () => {
    wrapper.setState({searchValue: 'The Office'})
    expect(wrapper.state('searchValue')).toEqual('The Office')
    wrapper.instance().resetInputs()
    expect(wrapper.state('searchValue')).toEqual('')
  })

  it('should have handleChange called on the search change', () => {
    wrapper.instance().handleChange = jest.fn()
    wrapper.find('.search-input').simulate('change')
    expect(wrapper.instance().handleChange).toHaveBeenCalled()
  })

  it('should set state for the searchValue', () => {
    expect(wrapper.state('searchValue')).toEqual('')
    wrapper.instance().handleChange(mockEvent)
    expect(wrapper.state('searchValue')).toEqual('The Office')
  })

  xit('should call setSearchShow', async () => {
    wrapper.instance().fetchWatchedShowId = jest.fn(() => 'The Office')
    wrapper.setState({searchValue: 'The Office'})
    let mockSetSearchShow = jest.fn()
    wrapper.setProps({setSearchShow: mockSetSearchShow})
    await wrapper.instance().handleSubmit(mockEvent)
    expect(mockSetSearchShow).toHaveBeenCalled()
  })

  xit('should set error state if could not fetch show', async () => {
    wrapper.instance().fetchWatchedShowId = jest.fn(() => "")
    expect(wrapper.state('error')).toEqual("")
    await wrapper.instance().handleSubmit(mockEvent)
    expect(wrapper.state('error')).toEqual("No results found. Please search again")
  })

  it('should call resetInputs', () => {
    wrapper.instance().fetchWatchedShowId = jest.fn(() => "The Office")
    wrapper.setState({searchValue: 'The Office'})
    wrapper.instance().resetInputs()
    expect(wrapper.state('searchValue')).toEqual('')
  })

})

describe("mapDispatchToProps", () => {
  it("calls dispatch with a setSearchShow action", () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setSearchShow({name: 'The Office'});
    const mappedDispatch = mapDispatchToProps(mockDispatch);
    mappedDispatch.setSearchShow({name: 'The Office'});
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });
});