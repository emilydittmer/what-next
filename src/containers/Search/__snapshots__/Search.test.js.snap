// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should match the snapshot 1`] = `ShallowWrapper {}`;

exports[`Search should render an error 1`] = `ShallowWrapper {}`;
